File: plugins/screenshot_timeline/helper/Sources/MagiVisionHelper/Permissions.swift
import Foundation
import CoreGraphics
import ApplicationServices

/// Returns the current Screen Recording TCC status as seen by THIS binary.
/// Apple's API only distinguishes granted vs not — there's no "not_determined"
/// surfacing through CGPreflight.
func checkScreenRecordingPermission() -> String {
    CGPreflightScreenCaptureAccess() ? "granted" : "denied"
}

/// Trigger the Screen Recording permission prompt for THIS binary.
/// No-op if already decided (macOS only prompts once per binary per session).
func requestScreenRecordingPermission() {
    _ = CGRequestScreenCaptureAccess()
}

/// Returns the current Accessibility TCC status as seen by THIS binary.
func checkAccessibilityPermission() -> String {
    AXIsProcessTrusted() ? "granted" : "denied"
}

/// Trigger the Accessibility permission prompt for THIS binary.
func requestAccessibilityPermission() {
    let promptKey = kAXTrustedCheckOptionPrompt.takeRetainedValue() as String
    let options = [promptKey: true] as CFDictionary
    _ = AXIsProcessTrustedWithOptions(options)
}

File: plugins/screenshot_timeline/permissions.py
"""macOS permission probes and request helpers for the screenshot_timeline plugin."""
from __future__ import annotations

import logging
from typing import Literal

logger = logging.getLogger(__name__)

PermissionStatus = Literal["granted", "denied", "not_determined", "unknown"]


def screen_recording_status() -> PermissionStatus:
    """Check Screen Recording permission without triggering a prompt."""
    try:
        from Quartz import CGPreflightScreenCaptureAccess  # type: ignore
    except Exception:
        logger.debug("permissions.screen_recording.import_failed")
        return "unknown"
    try:
        return "granted" if CGPreflightScreenCaptureAccess() else "denied"
    except Exception:
        logger.debug("permissions.screen_recording.query_failed")
        return "unknown"


def request_screen_recording() -> PermissionStatus:
    """Trigger the Screen Recording permission prompt (macOS shows it once per binary).

    Returns the immediate status. The actual user decision happens out-of-band; you should
    poll `screen_recording_status` afterwards.
    """
    try:
        from Quartz import CGRequestScreenCaptureAccess  # type: ignore
    except Exception:
        logger.debug("permissions.screen_recording_request.import_failed")
        return "unknown"
    try:
        CGRequestScreenCaptureAccess()
    except Exception:
        logger.debug("permissions.screen_recording_request.failed")
    return screen_recording_status()


def accessibility_status() -> PermissionStatus:
    """Check Accessibility permission without triggering a prompt."""
    try:
        from ApplicationServices import AXIsProcessTrusted  # type: ignore
    except Exception:
        logger.debug("permissions.accessibility.import_failed")
        return "unknown"
    try:
        return "granted" if AXIsProcessTrusted() else "denied"
    except Exception:
        return "unknown"


def request_accessibility() -> PermissionStatus:
    """Trigger the Accessibility permission prompt."""
    try:
        from ApplicationServices import (  # type: ignore
            AXIsProcessTrustedWithOptions,
            kAXTrustedCheckOptionPrompt,
        )
    except Exception:
        logger.debug("permissions.accessibility_request.import_failed")
        return "unknown"
    try:
        options = {kAXTrustedCheckOptionPrompt: True}
        AXIsProcessTrustedWithOptions(options)
    except Exception:
        logger.debug("permissions.accessibility_request.failed")
    return accessibility_status()


def all_statuses() -> dict[str, PermissionStatus]:
    return {
        "screen_recording": screen_recording_status(),
        "accessibility": accessibility_status(),
    }


__all__ = [
    "PermissionStatus",
    "screen_recording_status",
    "request_screen_recording",
    "accessibility_status",
    "request_accessibility",
    "all_statuses",
]
